File: server/image-editor.js
import path from "node:path";
import { fileURLToPath } from "node:url";
import { randomUUID } from "node:crypto";
import { mkdir, writeFile, unlink } from "node:fs/promises";
import { spawn } from "node:child_process";
import { qwenReferenceCandidates } from "./qwen-status.js";
import { QWEN_DRAFT } from "./qwen-image.js";

const SCRIPT = fileURLToPath(new URL("./image_editor.py", import.meta.url));
const ENGINE = "qwen-image-2.1";
const MODES = new Set(["edit", "style", "inpaint"]);

export function editorOptions(body) {
  const allowed = new Set(["action", "documentId", "source", "mode", "prompt", "refImages", "selection",
    "steps", "cfg", "seed", "refResolution", "transparent", "dit", "encoder", "vae"]);
  /* Fast draft is a Pictures setting. Masked edits are base-only, and edit and
   * style edits were not what the draft was judged on, so the editor says so
   * rather than calling the field unsupported. */
  if (body.draft !== undefined) throw new Error(QWEN_DRAFT.refusals.masked);
  const unsupported = Object.keys(body).filter(key => !allowed.has(key));
  if (unsupported.length) throw new Error(`Unsupported editor fields: ${unsupported.join(", ")}.`);
  const mode = body.mode ?? "edit";
  if (!MODES.has(mode)) throw new Error("mode must be edit, style or inpaint.");
  if (!!body.documentId === !!body.source) throw new Error("Choose exactly one documentId or source image filename.");
  const prompt = String(body.prompt || "").trim();
  if (!prompt || prompt.length > 8000) throw new Error("Describe the edit in 1–8000 characters.");
  const refImages = body.refImages ?? [];
  if (!Array.isArray(refImages) || refImages.length > 9 || refImages.some(n => typeof n !== "string" || !n.trim() || /[/\\]/.test(n)))
    throw new Error("Choose at most nine additional image filenames; the source is always image 1.");
  if (mode === "style" && !refImages.length) throw new Error("Style transfer needs at least one style reference (image 2).");
  if (mode === "inpaint" && refImages.length > 8) throw new Error("Masked edits allow eight extra references: image 1 is the source and image 2 is the selection mask.");
  if (mode === "inpaint" && (!body.selection || !Array.isArray(body.selection.shapes) || !body.selection.shapes.length))
    throw new Error("Draw a selection before requesting a masked edit.");
  const number = (key, fallback, min, max, integer = false) => {
    const value = body[key] ?? fallback;
    if (typeof value !== "number" || !Number.isFinite(value) || value < min || value > max || (integer && !Number.isInteger(value)))
      throw new Error(`${key} must be ${integer ? "an integer" : "a number"} from ${min} to ${max}.`);
    return value;
  };
  const resolution = number("refResolution", 1024, 0, 4096, true);
  if (resolution % 32) throw new Error("refResolution must be a multiple of 32, or 0 to preserve the source resolution.");
  if (body.transparent !== undefined && typeof body.transparent !== "boolean") throw new Error("transparent must be true or false.");
  if (mode === "inpaint" && body.transparent) throw new Error("Masked edits preserve source alpha outside the selection; transparent generation is available in Edit and Style modes.");
  const options = { engine: ENGINE, prompt, count: 1, refImages: [...refImages],
    steps: number("steps", 25, 1, 50, true), cfg: number("cfg", 1, 1, 10),
    seed: number("seed", Math.floor(Math.random() * 4294967296), 0, Number.MAX_SAFE_INTEGER, true),
    refSizing: "reference", refResolution: resolution, transparent: body.transparent ?? false,
    sampler: "euler", scheduler: "simple" };
  for (const key of ["dit", "encoder", "vae"]) {
    if (body[key] !== undefined) {
      if (typeof body[key] !== "string" || !body[key].endsWith(".safetensors") || /[/\\]/.test(body[key]))
        throw new Error(`${key} must be a native safetensors model filename.`);
      options[key] = body[key];
    }
  }
  return { mode, options };
}

/** All generation goes through the existing Image API/ArtRunner supplied by index. */
export function createImageEditor({ imageDir, inputDir, coverDir, python, generate, preflight = async () => {},
  register = async () => {}, documentChanged = async () => {}, runPython } = {}) {
  const jobs = new Map();
  let preparing = 0;
  const documentLocks = new Map();
  async function withDocumentLock(key, operation) {
    const previous = documentLocks.get(key) || Promise.resolve();
    let release;
    const held = new Promise(resolve => { release = resolve; });
    documentLocks.set(key, held);
    await previous;
    try { return await operation(); }
    finally { release(); if (documentLocks.get(key) === held) documentLocks.delete(key); }
  }
  const scratch = path.join(inputDir, "aiplay_editor");
  const run = runPython || (async (mode, payload) => {
    await mkdir(scratch, { recursive: true });
    const jobFile = path.join(scratch, `${randomUUID()}.json`);
    await writeFile(jobFile, JSON.stringify({ dir: imageDir, ...payload }));
    try {
      return await new Promise((resolve, reject) => {
        const proc = spawn(python, [SCRIPT, mode, jobFile], { windowsHide: true });
        let stdout = "", stderr = "";
        const timer = setTimeout(() => { proc.kill(); reject(new Error("The image editor CPU operation timed out.")); }, 120_000);
        proc.stdout.on("data", b => { stdout += b; });
        proc.stderr.on("data", b => { stderr = (stderr + b).slice(-4000); });
        proc.once("error", error => { clearTimeout(timer); reject(error); });
        proc.once("close", code => {
          clearTimeout(timer);
          try {
            const result = JSON.parse(stdout.trim().split(/\r?\n/).at(-1));
            if (code !== 0 || !result.ok) throw new Error(result.error || stderr || "Image operation failed.");
            resolve(result);
          } catch (error) { reject(new Error(error.message || stderr)); }
        });
      });
    } finally { await unlink(jobFile).catch(() => {}); }
  });
  const publicJob = job => ({ ok: true, id: job.id, status: job.status, mode: job.mode,
    documentId: job.documentId || null, source: job.source || null, seed: job.options.seed,
    referenceCount: job.options.refImages.length, width: job.width, height: job.height,
    coverage: job.coverage, createdAt: job.createdAt, candidate: job.candidate,
    sourcePreview: job.sourcePreview,
    error: job.error, warnings: job.warnings || [], runId: job.runId,
    semantics: job.mode === "inpaint" ? "Reference-guided generation, composited through the frozen selection. Zero-mask pixels are unchanged."
      : "Reference-guided edit. Accept adds a new layer and retains the original layers hidden." });
  async function preview(body) {
    if (!!body.id === !!body.doc) throw new Error("Choose one saved document id or document tree.");
    return run("preview", { documentId: body.id, doc: body.doc });
  }
  async function create(body, actor) {
    const { mode, options } = editorOptions(body);
    if (preparing + [...jobs.values()].filter(j => j.status === "generating").length >= 4)
      throw new Error("Four editor generations are already pending. Wait for one to finish.");
    ++preparing;
    try {
    // Keep the bounded review history without expiring an active generation.
    for (const [id, job] of jobs) if (job.status !== "generating" && Date.now() - job.createdAt > 86400000) jobs.delete(id);
    if (jobs.size >= 100) throw new Error("The editor review history is full; restart the app to clear completed session history.");
    await mkdir(inputDir, { recursive: true });
    await mkdir(scratch, { recursive: true });
    const id = randomUUID(), token = id.replaceAll("-", "").slice(0, 12);
    const sourceName = `aiplay_frame_${token}.png`, sourcePath = path.join(inputDir, sourceName);
    const maskPath = path.join(scratch, `${id}.npy`);
    const maskName = `aiplay_frame_${randomUUID().replaceAll("-", "").slice(0, 12)}.png`;
    const maskImagePath = path.join(inputDir, maskName);
    // Unless transparency is asked for, a reference with alpha is sent as
    // Qwen's vision tower sees it, over white. Its VAE would keep the alpha
    // and hand back a transparent generation.
    const references = options.transparent ? [] : options.refImages.map(name => ({ name,
      candidates: qwenReferenceCandidates(name, { inputDir, coverDir, imageDir }),
      out: path.join(inputDir, `aiplay_frame_${randomUUID().replaceAll("-", "").slice(0, 12)}.png`) }));
    const stagedOptions = { ...options,
      prompt: mode === "inpaint" ? `Edit <image 1> only where the selection mask in <image 2> is white. Black regions must stay unchanged. ${options.prompt}`
        : mode === "style" ? `Edit <image 1> using the visual style of <image 2>, preserving the subject and composition of <image 1>. ${options.prompt}` : options.prompt };
    let prepared;
    try {
      const { references: sent, ...frozen } = await run("prepare", { documentId: body.documentId, source: body.source,
        mode, selection: body.selection, out: sourcePath, maskOut: maskPath, maskImageOut: maskImagePath, references });
      prepared = frozen;
      stagedOptions.refImages = [sourceName, ...(mode === "inpaint" ? [maskName] : []), ...options.refImages.map((name, i) => sent?.[i] ?? name)];
      await preflight(stagedOptions);
    }
    catch (error) {
      await Promise.allSettled([sourcePath, maskPath, maskImagePath, ...references.map(r => r.out)].map(file => unlink(file)));
      throw error;
    }
    const job = { id, status: "generating", actor, createdAt: Date.now(), mode,
      options: stagedOptions, requestedPrompt: options.prompt, documentId: body.documentId, source: body.source,
      sourcePath, maskPath: mode === "inpaint" ? maskPath : null,
      ...prepared };
    jobs.set(id, job);
    // Deliberately detached from the HTTP response: UI and MCP poll the same job.
    job.done = (async () => {
      try {
        const generated = await generate(stagedOptions, actor);
        if (!generated?.name) throw new Error("Qwen did not return an image filename.");
        if (generated.seed != null) { job.options.seed = generated.seed; options.seed = generated.seed; }
        const name = `qwen_edit_${token}.png`;
        const { warnings = [], ...finished } = await run("finish", { sourcePath, generated: generated.name,
          maskPath: job.maskPath, out: path.join(imageDir, name), thumbOut: path.join(imageDir, `qwen_edit_${token}_t.png`) });
        if (warnings.length) job.warnings = [...(job.warnings || []), ...warnings];
        job.runId = generated.runId;
        await register(name, { engine: ENGINE, prompt: options.prompt, seed: options.seed,
          generatedFrom: generated.name, derivedFrom: body.source || null, documentId: body.documentId || null, operation: `qwen-${mode}`,
          refImages: options.refImages, masked: mode === "inpaint", runId: generated.runId }, actor);
        job.candidate = { name, url: `/api/image/${encodeURIComponent(name)}`, ...finished };
        job.status = "ready";
      } catch (error) { job.status = "error"; job.error = error.message; }
    })();
    return publicJob(job);
    } finally { --preparing; }
  }
  async function request(body, actor = "system") {
    const action = body.action || "create";
    if (action === "create") return create(body, actor);
    const job = jobs.get(body.id);
    if (!job) throw new Error("This editor job is unavailable or belongs to a previous app session.");
    if (action === "status") return publicJob(job);
    if (job.mutating) throw new Error("This editor result is already being changed.");
    if (action === "discard") {
      if (job.status !== "ready" && job.status !== "error") throw new Error("Only a finished, unaccepted edit can be discarded.");
      job.status = "discarded";
      return publicJob(job);
    }
    if (!["accept", "undo"].includes(action)) throw new Error("action must be create, status, accept, discard or undo.");
    if (action === "accept" && job.status !== "ready") throw new Error("Only a ready result can be accepted.");
    if (action === "undo" && job.status !== "accepted") throw new Error("Only an accepted result can be undone.");
    job.mutating = true;
    try {
      const result = await withDocumentLock(job.documentId || `source:${job.id}`, () => action === "accept" ? run("accept", {
        documentId: job.documentId, revision: job.revision, sourcePath: job.sourcePath,
        originalName: `qwen_original_${job.id.slice(0, 8)}.png`, title: job.source || "Qwen document edit",
        width: job.width, height: job.height, candidate: job.candidate.name,
        layerId: `qwen_${job.id.replaceAll("-", "")}`, layerName: `Qwen ${job.mode}: ${job.requestedPrompt.slice(0, 60)}`,
      }) : run("undo", { documentId: job.documentId, revision: job.acceptedRevision, before: job.before }));
      job.documentId = result.doc.id;
      if (action === "accept") { job.before = result.before; job.acceptedRevision = result.revision; }
      job.status = action === "accept" ? "accepted" : "undone";
      try { await documentChanged({ action, id: job.id, documentId: job.documentId, layerId: result.layerId,
        candidate: job.candidate.name }, actor); }
      catch (error) { (job.warnings ||= []).push(`Document saved, but its audit note failed: ${error.message}`); }
      return { ...publicJob(job), doc: result.doc, revision: result.revision, layerId: result.layerId };
    } finally { job.mutating = false; }
  }
  return { preview, request, paintTarget: body => run("paint-target", { doc: body.doc, ref: body.ref }),
    flattenReferences: references => run("flatten", { references }).then(result => result.references) };
}

File: server/image_editor.py
"""CPU-only preparation, document preview and exact masked Qwen composition.

Qwen Image 2.1 consumes references, not an inpainting mask. The selection is
resolved against a frozen source before generation, then applied here afterwards.
"""
import base64
import copy
import hashlib
import io
import json
import os
import sys

import numpy as np
from PIL import Image
import imgdoc
import imgselect


def revision(doc):
    return hashlib.sha256(json.dumps(doc, sort_keys=True, separators=(",", ":"),
                                     ensure_ascii=False).encode("utf-8")).hexdigest()


def library_path(directory, name):
    if not isinstance(name, str) or not name or name != os.path.basename(name) \
            or "/" in name or "\\" in name or name.startswith("."):
        raise ValueError("Use a library image filename, not a path.")
    path = os.path.join(directory, name)
    if not os.path.isfile(path):
        raise ValueError(f"Image is missing: {name}")
    return path


def open_doc(job):
    return imgdoc.store_job({"dir": job["dir"], "action": "open", "id": job["documentId"]})["doc"]


def render_doc(doc, directory):
    sources = {}
    def walk(layers):
        for layer in layers:
            for name in (layer.get("src"), (layer.get("mask") or {}).get("src")):
                if name:
                    sources[name] = library_path(directory, name)
            walk(layer.get("layers") or [])
    walk(doc.get("layers") or [])
    if int(doc.get("width", 0)) * int(doc.get("height", 0)) > 16_777_216:
        raise ValueError("The editor preview supports up to 16 megapixels.")
    report = {}
    pixels = imgdoc.render(doc, imgdoc.resolver_for(sources), 1, report)
    if report.get("missing"):
        raise ValueError("Document sources are missing: " + ", ".join(report["missing"]))
    return Image.fromarray(imgdoc.to_uint8(pixels), "RGBA"), report.get("warnings", [])


def preview(job):
    doc = open_doc(job) if job.get("documentId") else imgdoc.normalize(job.get("doc"))
    image, warnings = render_doc(doc, job["dir"])
    data = io.BytesIO()
    image.save(data, format="PNG")
    return {"dataUrl": "data:image/png;base64," + base64.b64encode(data.getvalue()).decode("ascii"),
            "width": image.width, "height": image.height, "revision": revision(doc), "warnings": warnings,
            "paintTargets": paint_targets(doc, job["dir"])}


PAINT_GUIDANCE = "Use Documents → Render & open composite to paint flattened pixels, or Qwen edit on the composed document."


def paint_targets(doc, directory):
    """Conservative eligibility; never pretend canvas coordinates are local.

    Reuse the renderer's affine calculation, including its implicit centering.
    Brushes and pixel-sampling selections need more than transformed endpoints
    for inverse mapping, so unsupported transforms are refused explicitly.
    """
    result = {}
    width, height = doc["width"], doc["height"]
    identity = np.array([[1., 0., 0.], [0., 1., 0.]])
    def walk(layers, ancestors):
        for layer in layers:
            chain = [*ancestors, layer]
            reason = None
            if layer.get("type") != "image" or not layer.get("src"):
                reason = "Only an image layer has source pixels to paint."
            size = None
            if reason is None:
                try:
                    with Image.open(library_path(directory, layer["src"])) as image:
                        size = image.size
                except (ValueError, OSError) as error:
                    reason = str(error)
            if reason is None and size != (width, height):
                reason = "The layer source is not the full canvas size; canvas strokes cannot be mapped to its pixels safely."
            if reason is None:
                for item in chain:
                    label = item.get("name") or item["id"]
                    if item.get("locked") or item.get("enabled") is False:
                        reason = f'Layer or group "{label}" is locked or hidden. Unlock and show it before painting.'
                        break
                    if any(isinstance(effect, dict) and effect.get("enabled") is not False for effect in item.get("effects") or []):
                        reason = f'Layer or group "{label}" has enabled effects; its displayed pixels may be displaced.'
                        break
                    matrix = imgdoc.interp.transform_matrix(item.get("transform") or {}, 0,
                        anchor_default=(width / 2, height / 2), position_default=(width / 2, height / 2))
                    if not np.allclose(matrix, identity, rtol=0, atol=1e-9):
                        reason = f'Layer or group "{label}" is transformed; canvas strokes cannot be mapped to its pixels safely.'
                        break
            result[layer["id"]] = {"ready": reason is None, "reason": (reason + " " + PAINT_GUIDANCE) if reason else "Full-canvas source with an identity pixel mapping."}
            walk(layer.get("layers") or [], chain)
    walk(doc.get("layers") or [], [])
    return result


def paint_target(job):
    doc = imgdoc.normalize(job["doc"])
    layer, _, _ = imgdoc.find_layer(doc, job["ref"])
    result = paint_targets(doc, job["dir"])[layer["id"]]
    if not result["ready"]:
        raise ValueError(result["reason"])
    return {**result, "ref": layer["id"]}


def flatten_references(references):
    """Qwen's vision tower sees a reference's alpha over white, but its VAE
    encodes all four channels, so one cutout reference turned a whole
    generation transparent. Hand Qwen the picture its vision tower sees.

    A name that is missing or unreadable stays as it is: staging refuses it
    with the message every other Qwen door gives."""
    names = []
    for reference in references:
        path = next((p for p in reference["candidates"] if os.path.isfile(p)), None)
        rgba = None
        if path:
            try:
                with Image.open(path) as image:
                    if any(band in ("A", "a") for band in image.getbands()) or "transparency" in image.info:
                        rgba = image.convert("RGBA")
            except OSError:
                pass
        if rgba is None or rgba.getextrema()[3][0] == 255:
            names.append(reference["name"])
            continue
        flat = Image.new("RGBA", rgba.size, (255, 255, 255, 255))
        flat.alpha_composite(rgba)
        flat.convert("RGB").save(reference["out"])
        names.append(os.path.basename(reference["out"]))
    return names


def flatten(job):
    """The same rule for a plain Qwen generation (stageQwenReferences)."""
    return {"references": flatten_references(job["references"])}


def prepare(job):
    doc = open_doc(job) if job.get("documentId") else None
    if doc:
        image, warnings = render_doc(doc, job["dir"])
    else:
        with Image.open(library_path(job["dir"], job["source"])) as src:
            image = src.convert("RGBA")
        warnings = []
    if image.width * image.height > 16_777_216:
        raise ValueError("Qwen edits support source canvases up to 16 megapixels.")
    image.save(job["out"])
    thumbnail = image.copy()
    thumbnail.thumbnail((512, 512), Image.Resampling.LANCZOS)
    preview_data = io.BytesIO()
    thumbnail.save(preview_data, format="PNG")
    coverage = None
    if job.get("mode") == "inpaint":
        selection = job.get("selection")
        if not isinstance(selection, dict) or not selection.get("shapes"):
            raise ValueError("Masked edit needs a nonempty selection.")
        notes = []
        mask = imgselect.resolve(selection, np.asarray(image, dtype=np.float32) / 255, notes)
        if notes:
            raise ValueError("Selection could not be resolved exactly: " + "; ".join(notes))
        if not np.any(mask > 0):
            raise ValueError("The selection covers no pixels.")
        # Float32 avoids turning a small feather value into a zero by quantisation.
        np.save(job["maskOut"], mask, allow_pickle=False)
        Image.fromarray(np.round(np.clip(mask, 0, 1) * 255).astype(np.uint8), "L").convert("RGB").save(job["maskImageOut"])
        coverage = float(np.mean(mask))
    return {"width": image.width, "height": image.height, "warnings": warnings,
            "document": doc, "revision": revision(doc) if doc else None, "coverage": coverage,
            "references": flatten_references(job.get("references") or []),
            "sourcePreview": "data:image/png;base64," + base64.b64encode(preview_data.getvalue()).decode("ascii")}


def over_source(generated, original):
    """The generation laid over the source; straight alpha, float32 0..1.

    A masked edit never asks for transparency, but Qwen can still return it.
    Pasted through the selection, that punched a hole in an opaque frame and
    showed the colour under the alpha. Opaque pixels are copied, so an opaque
    generation composites exactly as before; clear ones leave the source."""
    ag, ao = generated[..., 3:], original[..., 3:]
    alpha = ag + ao * (1 - ag)
    with np.errstate(divide="ignore", invalid="ignore"):
        rgb = (generated[..., :3] * ag + original[..., :3] * ao * (1 - ag)) / alpha
    over = np.concatenate([rgb, alpha], axis=-1)
    return np.where(ag >= 1, generated, np.where(ag <= 0, original, over))


def finish(job):
    with Image.open(job["sourcePath"]) as source:
        original = np.asarray(source.convert("RGBA")).copy()
    with Image.open(library_path(job["dir"], job["generated"])) as generated:
        image = generated.convert("RGBA")
    resized = image.size != (original.shape[1], original.shape[0])
    if resized:
        # Qwen produces 32-aligned dimensions. The edit always returns to the
        # exact document grid before applying its selection.
        image = image.resize((original.shape[1], original.shape[0]), Image.Resampling.LANCZOS)
    result = np.asarray(image).copy()
    warnings = []
    if job.get("maskPath"):
        mask = np.load(job["maskPath"], allow_pickle=False)
        if mask.shape != original.shape[:2]:
            raise ValueError("Frozen selection does not match the source dimensions.")
        generated = result.astype(np.float32) / 255
        selected = mask > 0
        clear = selected & (generated[..., 3] < .5)
        if not np.any(selected & ~clear):
            raise ValueError("Qwen returned the whole selection transparent, so the edit would change nothing. "
                             "A reference or source with transparency can cause this.")
        share = float(np.sum(mask[clear]) / np.sum(mask[selected]))
        if share >= .01:
            warnings.append(f"Qwen returned {share:.0%} of the selection transparent; those pixels keep the source.")
        source01 = original.astype(np.float32) / 255
        result = imgdoc.to_uint8(imgselect.blend(source01, over_source(generated, source01), mask))
        result[mask <= 0] = original[mask <= 0]  # includes hidden RGB and alpha
    final = Image.fromarray(result, "RGBA")
    final.save(job["out"])
    thumb = final.copy()
    thumb.thumbnail((256, 256), Image.Resampling.LANCZOS)
    thumb.save(job["thumbOut"])
    return {"width": final.width, "height": final.height, "resizedToSource": resized, "warnings": warnings}


def _accept_locked(job):
    if job.get("documentId"):
        before = open_doc(job)
        if revision(before) != job["revision"]:
            raise ValueError("The document changed after this edit started. Keep the candidate, or generate again from the current document.")
    else:
        # Keep the frozen original, even if someone replaced the source file
        # while the GPU was running. It becomes an ordinary library layer.
        with Image.open(job["sourcePath"]) as image:
            image.save(os.path.join(job["dir"], job["originalName"]))
        before = imgdoc.normalize({"name": job.get("title") or "Qwen edit",
            "width": job["width"], "height": job["height"],
            "layers": [{"id": "original", "type": "image", "name": "Original", "src": job["originalName"]}]})
    doc = copy.deepcopy(before)
    if float(imgdoc._rgba01(before.get("bg"), (0., 0., 0., 0.))[3]) > 0:
        # Keep the old background in the saved layer tree, not just in this
        # session's undo snapshot. A native solid needs no extra raster asset.
        doc, _ = imgdoc.apply_edits(doc, [{"op": "add_layer", "index": 0, "layer": {
            "type": "solid", "name": "Original document background", "enabled": False,
            "size": [doc["width"], doc["height"]], "color": copy.deepcopy(before["bg"]),
        }}])
    # The result already includes the old document background. Keeping it
    # beneath the replacement would fill new transparency or double its alpha.
    doc["bg"] = [0, 0, 0, 0]
    for layer in doc["layers"]:
        layer["enabled"] = False
    doc, _ = imgdoc.apply_edits(doc, [{"op": "add_layer", "layer": {
        "id": job["layerId"], "type": "image", "name": job["layerName"], "src": job["candidate"]}}])
    saved = imgdoc.store_job({"dir": job["dir"], "action": "save", "doc": doc})
    current = imgdoc.store_job({"dir": job["dir"], "action": "open", "id": saved["id"]})["doc"]
    return {"doc": current, "revision": revision(current), "before": before, "layerId": job["layerId"]}


def _undo_locked(job):
    current = open_doc(job)
    if revision(current) != job["revision"]:
        raise ValueError("The document has newer edits. Undo those first, or toggle the Qwen layer and original layers manually.")
    before = dict(job["before"], id=current["id"], slug=current["slug"])
    saved = imgdoc.store_job({"dir": job["dir"], "action": "save", "doc": before})
    restored = imgdoc.store_job({"dir": job["dir"], "action": "open", "id": saved["id"]})["doc"]
    return {"doc": restored, "revision": revision(restored)}


def accept(job):
    with imgdoc.shelf_lock(job):
        return _accept_locked(job)


def undo(job):
    with imgdoc.shelf_lock(job):
        return _undo_locked(job)


def main():
    with open(sys.argv[2], encoding="utf-8") as handle:
        job = json.load(handle)
    try:
        action = {"preview": preview, "paint-target": paint_target, "prepare": prepare, "finish": finish,
                  "accept": accept, "undo": undo, "flatten": flatten}[sys.argv[1]]
        print(json.dumps({"ok": True, **action(job)}))
    except Exception as error:
        print(json.dumps({"ok": False, "error": str(error)}))
        sys.exit(1)


if __name__ == "__main__":
    main()
